### PR TITLE
[FLINK-21891][table-runtime-blink] Use FileSystem instance to delete the temporary directory '.staging_xxx'

### PR DESCRIPTION
**Pull request by itinycheng**
## What is the purpose of the change

This PR fixes the bug that the temporary directory '.staging_xxx' isn't deleted after writing data to hive finished in Flink SQL batch mode.

## Brief change log

  - Use the `FileSystem` instance created by `fsFactory` to delete temporary files, because the type of `FileSystem` is different in different contexts.

## Verifying this change

- This change is already covered by existing tests, such as *org.apache.flink.table.filesystem.FileSystemOutputFormatTest*, and I have done an integration test in the HDFS environment. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduces a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Review by KurtYoung — COMMENTED** [review]
Is it possible to add a test case? To make sure we don't reintroduce such bug in the future.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/FileSystemOutputFormat.java [hunk 1]
```diff
@@ -85,15 +85,26 @@ private FileSystemOutputFormat(
...
                             fsFactory, msFactory, overwrite, tmpPath, partitionColumns.length);
             committer.commitUpToCheckpoint(CHECKPOINT_ID);
         } catch (Exception e) {
-            throw new TableException("Exception in finalizeGlobal", e);
-        } finally {
-            new File(tmpPath.getPath()).delete();
+            exception = new TableException("Exception in finalizeGlobal", e);
+        }
+
+        // delete the temporary directory
```
Comment (lirui-apache) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/FileSystemOutputFormat.java:98]: Why not keep this in the finally block?

**Comment by lirui-apache** [comment]
The issue only happens for non-local FS. So if we want to add a test case, I think we should use `MiniDFSCluster` for it.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/FileSystemOutputFormat.java (hunk 1 above)
Comment (itinycheng) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/FileSystemOutputFormat.java:98]: I think when an exception is raised in both `try` and `finally` blocks, the exception in the try block will be lost.

**Comment by itinycheng** [comment]
> The issue only happens for non-local FS. So if we want to add a test case, I think we should use `MiniDFSCluster` for it.

Agree.

**Comment by itinycheng** [comment]
> The issue only happens for non-local FS. So if we want to add a test case, I think we should use `MiniDFSCluster` for it.

Yes, and write a hive test case in `MiniDFSCluster`?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/FileSystemOutputFormat.java (hunk 1 above)
Comment (lirui-apache) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/FileSystemOutputFormat.java:98]: I guess we can just catch and discard the exceptions during removing the temp folder (with a warning log perhaps). There's no need to fail the job for it.

**Comment by lirui-apache** [comment]
> > The issue only happens for non-local FS. So if we want to add a test case, I think we should use `MiniDFSCluster` for it.
> 
> Yes, and write a hive test case in `YARNApplicationITCase`, right?

No, `YARNApplicationITCase` shouldn't depend on hive. We can add the test case in `flink-table-runtime-blink` or `flink-connector-hive`.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/FileSystemOutputFormat.java (hunk 1 above)
Comment (itinycheng) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/FileSystemOutputFormat.java]: I get it, thanks.

**Comment by itinycheng** [comment]
Solved by https://github.com/apache/flink/pull/15541

**Pull request closed without merging**